### ISSUE TITLE
Contract instantiate produces OutputBufferTooSmall error

### ISSUE THREAD

**Issue by lana-shanghai** [polkascan/py-substrate-interface#56]
Hi Arjan, it's probably a substrate question, not the interface, but I would like to know what causes this OutputBufferTooSmall error when I try to run 
```
call = substrate.compose_call(
    call_module='Contracts',
    call_function='instantiate',
    call_params={
        'endowment': 1000000,
        'gas_limit': 1000000000,
        'code_hash': code_hash,
        'data': '0x{}'.format(code_bytes.hex())
    }
)
```
is it something about the data? What goes into data, I didn't find in the docs. Not the contract code I assume, as it already was passed with the contracts.put_code ..?

<img width="539" alt="Screenshot 2020-11-24 at 13 06 21" src="https://user-images.githubusercontent.com/31368580/100092386-4125e600-2e56-11eb-8dfb-246935cc9be3.png">

**Comment by arjanz** [comment]
Ok this is how far I could go with current functionality, in the code example below I successfully uploaded and deployed a contract. But the big caveat here is that the Python library doesn't (yet) support the encoding/determining how to interact with a contract. This will be a project on its own to make usable interface functions in the Python library for this, most likely having to parse the generated metadata ABI JSON file when creating the contract. 

In my code example I copy paste the initial state data following the tutorial https://substrate.dev/substrate-contracts-workshop/ and figuring out what is sent by connecting a local hosting version of https://polkadot.js.org/apps/ to the local node.

```python
import os
from hashlib import blake2b

from substrateinterface import SubstrateInterface, SubstrateRequestException, Keypair, ss58_encode

substrate = SubstrateInterface(
    url="ws://127.0.0.1:9944",
    type_registry_preset='substrate-node-template'
)

keypair = Keypair.create_from_uri('//Alice', address_type=42)

path = os.path.join(os.path.dirname(__file__), 'flipper.wasm')
with open(os.path.abspath(path), 'rb') as fp:
    code_bytes = fp.read()

code_hash = '0x{}'.format(blake2b(code_bytes, digest_size=32).digest().hex())

call = substrate.compose_call(
    call_module='Contracts',
    call_function='put_code',
    call_params={
        'code': '0x{}'.format(code_bytes.hex())
    }
)

extrinsic = substrate.create_signed_extrinsic(call=call, keypair=keypair)

try:
    result = substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)

    print('Contract uploaded, extrinsic hash: {}'.format(
        result['extrinsic_hash']
    ))

    # Fetch code hash in events
    events = substrate.get_runtime_events(result['block_hash'])

    for event in events['result']:
        if event['event_id'] == 'CodeStored':
            event_code_hash = events['result'][1]['params'][0]['value']

            # just to prove that the code_hash is deterministic
            assert event_code_hash == code_hash

    # Instantiate contract
    call = substrate.compose_call(
        call_module='Contracts',
        call_function='instantiate',
        call_params={
            'endowment': 1000000000000000,
            'gas_limit': 1000000000000,
            'code_hash': code_hash,
            'data': '0x6a3712e2'  # -> Copied from result using Canvas tutorial, determining the initial state and interacting with contracts is not yet supported
        }
    )

    extrinsic = substrate.create_signed_extrinsic(call=call, keypair=keypair)

    result = substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)

    print('Contract deployed, extrinsic hash: {}'.format(
        result['extrinsic_hash']
    ))

    # Get contract address
    events = substrate.get_runtime_events(result['block_hash'])

    for event in events['result']:
        if event['event_id'] == 'Instantiated':
            contract_address = event['params'][1]['value']
            print("Contract address: {}".format(ss58_encode(contract_address)))

except SubstrateRequestException as e:
    print("Failed to send: {}".format(e))
```

**Comment by lana-shanghai** [comment]
Oh I see, so I'd need this snippet from the metadata file, right? 

```
 "docs": [
          " Creates a new ERC-20 contract with the specified initial supply."
        ],
        "name": [
          "new"
        ],
        "selector": "0xd183512b"

**Comment by arjanz** [comment]
Yes you're right, I used the 'default' constructor which has a selector of `0x6a3712e2`. So you could basically get it from the json file.. Then probably you can use `0xc096a5f3` as a contract call to flip the stored `bool`, but not yet sure how to pass arguments to a contract call. Probably scale-encoded according to the metadata, but didn't really looked into this to be honest. Any help would of course be appreciated :)

**Comment by arjanz** [comment]
![Screenshot 2020-11-25 at 14 16 39](https://user-images.githubusercontent.com/5286904/100232622-e14c3f80-2f28-11eb-8b7f-d0d2c207368f.png)

Confirmed to use `0xc096a5f3` as data to call the `flip` function

**Comment by lana-shanghai** [comment]
Actually, I did this with the previous version of ink, you chain them together in hex, but unfortunately some of the values had some big endian encoding, and had to be flipped backwards. See example of call here https://github.com/gridsingularity/d3a/blob/f7093b30b44c6b7b7a67d40bea918430f44e7eaa/src/d3a/models/market/blockchain_interface.py#L96 

and the helper function definitions here: 
https://github.com/gridsingularity/d3a/blob/f7093b30b44c6b7b7a67d40bea918430f44e7eaa/src/d3a/models/market/blockchain_utils.py#L49

I think I could add a PR with a general parser, to be able to call the functions just by their name, and add the params depending on their type

**Comment by lana-shanghai** [comment]
And yes, it took me HOURS to figure this out, because the UI didn't print out this data and chopped it off. Now I think you can copy it, at least they made an enhancement to be able to copy the hash from any page in the UI. I'll check the data field :)

**Comment by arjanz** [comment]
Seems like you entered the wondrous world of SCALE encoding/decoding :) You should have a look at `encode_scale` https://github.com/polkascan/py-substrate-interface/blob/master/substrateinterface/__init__.py#L2032 because that is most likely how you encode the arguments for a contract call (like `bool` or `u16` or `compact<balance>`) in the correct format. I do the same for encoding the parameter for a call in an extrinsic for example.

**Comment by lana-shanghai** [comment]
Hey Arjan, I've found a new bug/feature, hehe, so basically I have been able to deploy the same contract numerous times only if I add a random byte to the end, which is I guess equivalent to nonce in ethereum... which is why I wanted to ask how to fetch my wallet addresses's nonce? Not to make the process random, but deterministic. The 02 at the end of data string is my random byte. 

```
# Instantiate contract
    call = substrate.compose_call(
        call_module='Contracts',
        call_function='instantiate',
        call_params={
            'endowment': 1000000000000,
            'gas_limit': 1000000000000,
            'code_hash': code_hash,
            'data': '0xd183512b02'
        }
    )
```

**Comment by arjanz** [comment]
You can use the function `substrate.get_account_nonce("5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY")` to retrieve the nonce.

Interesting about the random bytes, I came across the 'deployment salt' option in PolkadotJs Apps but now you mention this I realize where that is for..

<img width="885" alt="Screenshot 2020-11-30 at 18 46 50" src="https://user-images.githubusercontent.com/5286904/100645377-e05a4a00-333c-11eb-9d72-8cdaf7660f33.png">
<img width="1043" alt="Screenshot 2020-11-30 at 18 48 08" src="https://user-images.githubusercontent.com/5286904/100645386-e3553a80-333c-11eb-9401-d8747915b904.png">


Also interesting for you is that in december I will reserve some time to make better interfacing classes for handling with ink!3 contracts. I would like to discuss some requirements and best practices on your end, are on Element/Riot by any chance? (@arjanz:matrix.org)

**Comment by wzli** [comment]
Hi, I wanted to chime in for additional interest in being able to interact with ink contracts via python. From this discussion, as far as I can tell ATM things should work as long as the data field in the call param of the call extrinsic is correctly formatted? And the spec is (correct me if I'm wrong): selector + param + param + ... + param + salt, encoded in SCALE. I'd be great to have a nice interface generated by parsing the contract's metadata json but building the block manually isn't too bad.

I've also tested deploying multiple of the same contracts, and it seem to work for me when I change the input param during deploy without a salt. Maybe the contract id is just blake2b of account + code_hash + data.

**Comment by arjanz** [comment]
I just released an initial version of ink! contract interfacing classes, still heavily under construction, but you can check out an example here:

https://github.com/polkascan/py-substrate-interface#ink-contract-interfacing-work-in-progress

**Comment by wzli** [comment]
Gave it a quick try, ran into some issues.
I suppose the intended usage is first
from substrateinterface.contracts import * 
it fails saying module "exception" not found. I got it got work by changing import exception to import substrateinterface.exception in contracts.py

Upload wasm worked.

When I try to instantiate a contract, I get the following error:
`  File "/home/wzli/.local/lib/python3.6/site-packages/substrateinterface/contracts.py", line 162, in deploy
    data = self.metadata.generate_constructor_data(name=constructor, args=args)
  File "/home/wzli/.local/lib/python3.6/site-packages/substrateinterface/contracts.py", line 71, in generate_constructor_data
    value=args[arg['name']]
  File "/home/wzli/.local/lib/python3.6/site-packages/substrateinterface/__init__.py", line 2167, in encode_scale
    type_string=type_string, metadata=self.metadata_decoder, runtime_config=self.runtime_config
  File "/home/wzli/.local/lib/python3.6/site-packages/scalecodec/base.py", line 381, in get_decoder_class
    raise NotImplementedError('Decoder class for "{}" not found'.format(type_string))
NotImplementedError: Decoder class for "str" not found
`
Seems like it can't handle when the metadata contains string fields?

Also another question is, how would one subscribe or look up event's generated from specific contract instances?

**Comment by arjanz** [comment]
> Gave it a quick try, ran into some issues.
> I suppose the intended usage is first
> from substrateinterface.contracts import *
> it fails saying module "exception" not found. I got it got work by changing import exception to import substrateinterface.exception in contracts.py

Hm I see there are some circular reference issues, I will relocate most code out of `__init__.py` to avoid this

> When I try to instantiate a contract, I get the following error:
> ` File "/home/wzli/.local/lib/python3.6/site-packages/substrateinterface/contracts.py", line 162, in deploy data = self.metadata.generate_constructor_data(name=constructor, args=args) File "/home/wzli/.local/lib/python3.6/site-packages/substrateinterface/contracts.py", line 71, in generate_constructor_data value=args[arg['name']] File "/home/wzli/.local/lib/python3.6/site-packages/substrateinterface/__init__.py", line 2167, in encode_scale type_string=type_string, metadata=self.metadata_decoder, runtime_config=self.runtime_config File "/home/wzli/.local/lib/python3.6/site-packages/scalecodec/base.py", line 381, in get_decoder_class raise NotImplementedError('Decoder class for "{}" not found'.format(type_string)) NotImplementedError: Decoder class for "str" not found`
> Seems like it can't handle when the metadata contains string fields? Maybe I'm passing in string arguments wrong?

I'm still working on a complete implementation of the contract type registry, so our scalecodec knows how to deal with the introduced types. Support is still limited and I only tested it with the ERC20 contract, I hope to have a more complete version somewhere this week.

> Also another question is, how would one subscribe or look up event's generated from specific contract instances?

Support for subscriptions are limited atm because this is a using a sync implementation of websockets. When contract extrinsics are submitted, the library will block until the extrinsic is included in a block and retrieves the triggered events from storage (which will be exposed in the returned `ContractReceipt` object). A todo is to parse the triggered `ContractExecution` event which will give more  information about the specific contract events.

**Comment by arjanz** [comment]
@wzli is it possible for you to share your `metadata.json` file? Then I have some more example data to run my tests with

**Comment by wzli** [comment]
> @wzli is it possible for you to share your `metadata.json` file? Then I have some more example data to run my tests with

Sure, I'm testing an auction contract for off chain assets:

[metadata.json.txt](https://github.com/polkascan/py-substrate-interface/files/5698396/metadata.json.txt)

**Comment by arjanz** [comment]
Just released a version with a more complete metadata parsing, so I think your contract can also be parsed now, except for the `str` type. In Substrate `str` are usually represented as `Vec<u8>`, because `str` is not fixed length and cannot be SCALE decoded.

**Comment by wzli** [comment]
> Just released a version with a more complete metadata parsing, so I think your contract can also be parsed now, except for the `str` type. In Substrate `str` are usually represented as `Vec<u8>`, because `str` is not fixed length and cannot be SCALE decoded.

Okay, thanks, it's not an issue for me to change the type to something else equivalent but just to clarify, as far as I am aware Vec in rust is not "fixed length" as it can dynamically grow as required hence no static storage layout, did you mean byte array instead?
In ink contracts, std containers such as std::Vec and std::String are not supported anyway, and you're supposed to use the ones provided in the ink_prelude crate as swap in replacements, so I assumed that there would be native serialization given the types are custom.
Namely, between the two I chose string due to unicode support but that's not a requirement:
https://paritytech.github.io/ink/ink_prelude/string/index.html
https://paritytech.github.io/ink/ink_prelude/vec/index.html

 So you're saying use ink_prelude::Vec<u8> to store string data in the contract?

**Comment by arjanz** [comment]
> Okay, thanks, it's not an issue for me to change the type to something else equivalent but just to clarify, as far as I am aware Vec in rust is not "fixed length" as it can dynamically grow as required hence no static storage layout, did you mean byte array instead?

I actually meant predictable length, because SCALE has no delimiter or context, the `vec<u8>` is serialised as the element count `compact<u32>` and then followed by the elements, so the deserialiser can determine in advance how many bytes should be used to deserialise the type. 

> In ink contracts, std containers such as std::Vec and std::String are not supported anyway, and you're supposed to use the ones provided in the ink_prelude crate as swap in replacements, so I assumed that there would be native serialization given the types are custom provided.
> Namely, between the two I chose string due to unicode support but that's not a requirement:
> https://paritytech.github.io/ink/ink_prelude/string/index.html
> https://paritytech.github.io/ink/ink_prelude/vec/index.html
> So you're saying use ink_prelude::Vec to store string data in the contract?


I could be mistaken and the `str` is a new introduced type in ink that behaves in the same way, I haven't dived into these special types yet, but I'll give it a go tomorrow.

**Comment by arjanz** [comment]
@wzli You are right, the `str` type introduced in ink actually maps to a `Vec<u8>` and can be (de)serialized without a problem. I added a mapping in the type registry for this and run a test contract, works like a charm. 

Are you able to read and exec the messages of your contract now?

**Comment by wzli** [comment]
> @wzli You are right, the `str` type introduced in ink actually maps to a `Vec<u8>` and can be (de)serialized without a problem. I added a mapping in the type registry for this and run a test contract, works like a charm.
> 
> I just released v0.11.4, are you able to read and exec the messages of your contract now?
Will check after the holidays, thanks.
